src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -563,5 +623,48 @@ void RayletBasedActorScheduler::HandleWorkerLeaseReply(
   }
 }
 
+void GcsActorScheduler::HandleWorkerLeaseRejectedReply(
+    std::shared_ptr<GcsActor> actor, const rpc::RequestWorkerLeaseReply &reply) {
+  // The request was rejected because of insufficient resources.
+  if (actor->GetScheduledByGcs()) {
+    auto node_id = actor->GetNodeID();
+    auto &cluster_resource_manager =
+        cluster_task_manager_->GetClusterResourceScheduler()->GetClusterResourceManager();
+    auto acquired_resources = ResourceMapToResourceRequest(
+        actor->GetCreationTaskSpecification().GetRequiredResources().GetResourceMap(),
+        false);
+    cluster_resource_manager.AddNodeAvailableResources(
+        scheduling::NodeID(node_id.Binary()), acquired_resources);
+    actor->UpdateAddress(rpc::Address());
+    if (normal_task_resources_changed_callback_) {
+      normal_task_resources_changed_callback_(node_id, reply.resources_data());
+    }
+  }
+  Reschedule(actor);
+}
+
+void GcsActorScheduler::OnActorDestruction(std::shared_ptr<GcsActor> actor) {
+  if (actor->GetScheduledByGcs()) {
+    auto &cluster_resource_manager =
+        cluster_task_manager_->GetClusterResourceScheduler()->GetClusterResourceManager();
+    auto acquired_resources = ResourceMapToResourceRequest(
+        actor->GetCreationTaskSpecification().GetRequiredResources().GetResourceMap(),
+        false);
+    cluster_resource_manager.AddNodeAvailableResources(
+        scheduling::NodeID(actor->GetNodeID().Binary()), acquired_resources);
+    cluster_task_manager_->ScheduleAndDispatchTasks();
```

Comment:
should we call `normal_task_resources_changed_callback_` here? also should we create a helper function if this block is same as `HandleWorkerLeaseRejectedReply` ?

Reply:
> should we call `normal_task_resources_changed_callback_` here? 

The actor's resources have nothing to do with normal_task_resources, so we don't need to call the callback.

> also should we create a helper function if this block is same as `HandleWorkerLeaseRejectedReply` ?

Sure, just did.